Sources/Changeset.swift

```diff
@@ -0,0 +1,100 @@
+import Foundation
+
+/// Represents an atomic batch of changes made to a collection.
+///
+/// A `Changeset` represents changes as **offsets** of elements. You may
+/// subscript a collection of zero-based indexing with the offsets, e.g. `Array`. You must
+/// otherwise convert the offsets into indices before subscripting.
+public struct Changeset {
+	/// Represents the context of a move operation applied to a collection.
+	public struct Move {
+		public let source: Int
+		public let destination: Int
+		public let isMutated: Bool
+
+		public init(source: Int, destination: Int, isMutated: Bool) {
+			(self.source, self.destination, self.isMutated) = (source, destination, isMutated)
+		}
+	}
+
+	/// The offsets of inserted elements **after** the removals were applied.
+	///
+	/// - important: To obtain the actual index, you must apply
+	///              `Collection.index(self:_:offsetBy:)` on the current snapshot, the
+	///              start index and the offset.
+	public var inserts = IndexSet()
+
+	/// The offsets of removed elements **prior to** any changes being applied.
+	///
+	/// - important: To obtain the actual index, you must apply
+	///              `Collection.index(self:_:offsetBy:)` on the previous snapshot, the
+	///              start index and the offset.
+	public var removals = IndexSet()
+
+	/// The offsets of position-invariant mutations.
+	///
+	/// `mutations` only implies an invariant relative position. The actual indexes can
+	/// be different, depending on the collection type.
+	///
+	/// If an element has both changed and moved, it would be included in `moves` with an
+	/// asserted mutation flag.
+	///
+	/// - important: To obtain the actual index, you must apply
+	///              `Collection.index(self:_:offsetBy:)` on the relevant snapshot, the
+	///              start index and the offset.
+	public var mutations = IndexSet()
+
+	/// The offset pairs of moves with a mutation flag as the associated value.
+	///
+	/// - important: To obtain the actual index, you must apply
+	///              `Collection.index(self:_:offsetBy:)` on the relevant snapshot, the
+	///              start index and the offset.
+	public var moves = [Move]()
```

Comment:
What's the order of inserts, removals, mutations, and moves? You've documented that inserts happen after removals, but not when moves and mutations occur.